[WebGPU] HDR should use extended color space

mwyrzykowski opened the pull request:
#### 693449d75e78899b726458c5ab5aace05a89dbfb
<pre>
[WebGPU] HDR should use extended color space
<a href="https://bugs.webkit.org/show_bug.cgi?id=292577">https://bugs.webkit.org/show_bug.cgi?id=292577</a>
<a href="https://rdar.apple.com/150724620">rdar://150724620</a>

Reviewed by Gerald Squelart.

For extended tone mapping mode, we should also use an
extended color space otherwise we won&apos;t get the correct
range of color values.

* Source/WebCore/html/canvas/GPUCanvasContextCocoa.mm:
(WebCore::toWebCoreColorSpace):
(WebCore::GPUCanvasContextCocoa::configure):
(WebCore::GPUCanvasContextCocoa::colorSpace const):
* Source/WebCore/platform/graphics/DestinationColorSpace.cpp:
(WebCore::DestinationColorSpace::ExtendedDisplayP3):
* Source/WebCore/platform/graphics/DestinationColorSpace.h:

Canonical link: <a href="https://commits.webkit.org/294545@main">https://commits.webkit.org/294545@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8b6b9332c2a99e89cc9a3442d8ed7519fa569e24

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102248 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21916 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12231 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107407 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/52884 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/104287 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22227 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30423 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77800 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/34791 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105254 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17201 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92301 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58138 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17032 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10329 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52242 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86861 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10399 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/109783 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29380 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/21655 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/86782 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/29743 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88504 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86370 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21967 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31178 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8896 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/23622 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29308 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/34603 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29119 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32442 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/30678 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

squelart commented on Source/WebCore/html/canvas/GPUCanvasContextCocoa.mm:
```diff
@@ -248,13 +248,17 @@ static bool equalConfigurations(const auto& a, const auto& b)
+#if ENABLE(PREDEFINED_COLOR_SPACE_DISPLAY_P3)
+        return toneMapping.mode == GPUCanvasToneMappingMode::Standard ? DestinationColorSpace::DisplayP3() : DestinationColorSpace::ExtendedDisplayP3();
+#else
+        return DestinationColorSpace::SRGB();
```
Shouldn't you handle extended in this case as well?

mwyrzykowski replied:
yes good idea, it's 1 on Mac and iOS family but I think it is 0 on Mac Catalyst

squelart reviewed (approved):
r+ with nit (unless you disagree with it of course).
Thank you for this fix!

mwyrzykowski committed:
[WebGPU] HDR should use extended color space
https://bugs.webkit.org/show_bug.cgi?id=292577
rdar://150724620

Reviewed by Gerald Squelart.

For extended tone mapping mode, we should also use an
extended color space otherwise we won't get the correct
range of color values.

* Source/WebCore/html/canvas/GPUCanvasContextCocoa.mm:
(WebCore::toWebCoreColorSpace):
(WebCore::GPUCanvasContextCocoa::configure):
(WebCore::GPUCanvasContextCocoa::colorSpace const):
* Source/WebCore/platform/graphics/DestinationColorSpace.cpp:
(WebCore::DestinationColorSpace::ExtendedDisplayP3):
* Source/WebCore/platform/graphics/DestinationColorSpace.h:

Canonical link: https://commits.webkit.org/294545@main
